File: android/src/main/java/com/bluesnapsdkreactnative/utils/AwaitLock.kt
package com.bluesnapsdkreactnative.utils

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking



class AwaitLock() {
  private var isLocked: Boolean = false
  private var result: Any? = null
  private lateinit var timeOut: Job;


  suspend fun startLock(timeoutSeconds: Long?) = CoroutineScope(Dispatchers.Main).launch {
    isLocked = true;
    if (timeoutSeconds != null && timeoutSeconds != 0L) {
      timeOut = launch (Dispatchers.Default) {
        delay(timeoutSeconds * 1000L)
        isLocked = false
      }
    }
  }

  suspend fun <T> stopLock(withResult: T?) {
    this.isLocked = false;
    this.timeOut.cancelAndJoin()

    if(withResult != null) {
      this.result = withResult
    }
  }

  tailrec suspend fun awaitLock(): Any? {
    delay(1000)
    return if (isLocked) {
      awaitLock()
    } else {
      result
    }
  }

}

File: ios/AwaitLock.swift
//
//  AwaitLock.swift
//  BluesnapSdkReactNative
//
//  Created by MacOS on 7/12/23.
//  Copyright © 2023 Facebook. All rights reserved.
//

import Foundation
import PassKit

@available(iOS 13.0, *)
class AwaitLock {
    
    internal var isLocked: Bool = false
    internal var result: Any?
    internal var timeoutTask: Task<(), Never>?
    
    init() {
        self.isLocked = false;
        self.result = nil;
    }
    
    @available(iOS 13.0.0, *)
    func startLock(timeoutSeconds: Double?) async {
        self.isLocked = true;
        
        //Auto released lock after timeout
        if (timeoutSeconds != nil && timeoutSeconds != 0) {
            
            timeoutTask = Task {
                do {
                    try await Task.sleep(seconds: timeoutSeconds!);
                    self.isLocked = false;
                } catch {}
            }
        }
        
    }
    
    func stopLock<T: AnyObject>(withResult: T?) {
        self.isLocked = false;
        self.timeoutTask?.cancel();
        
        if(withResult != nil) {
            self.result = withResult!;
        }
    }
    
    @available(iOS 13.0.0, *)
    func awaitLock() async -> Any? {
        do {
            try await Task.sleep(seconds: 1);
            
            if (isLocked) {
                return await self.awaitLock();
            } else {
                return self.result;
            }
            
        } catch {
            return nil;
        }
        
    }
}

@available(iOS 13.0.0, *)
extension Task where Success == Never, Failure == Never {
    static func sleep(seconds: Double) async throws {
        let duration = UInt64(seconds * 1_000_000_000);
        try await Task.sleep(nanoseconds: duration);
    }
}
